Fixes #11095 - IllegalStateException for completed requests with GzipHandler

sbordet opened the pull request:
Fixed 2 problems:
* In ServletChannel, in COMPLETE state, there was a check comparing the number of bytes written with that declared by Content-Length. Unfortunately the check was wrong in case of gzip, because it was comparing the application length with the gzipped length, resulting in a late sendErrorOrAbort() that was an abort() because the response was already fully sent. Furthermore, if sendErrorOrAbort() was actually an abort(), there was an unnecessary attempt to complete the output.
* In GzipHandlerResponseAndCallback there were 2 last writes: one when the application calls output.close(), and one when the callback is succeeded. Furthermore, when no content needs to be written, it was still compressed, causing an IOException in ChannelResponse (again when double-checking the bytes written, since Content-Length was 0, but the gzip header and trailer were written).

sbordet committed:
Fixes #11095 - IllegalStateException for completed requests wih GzipHandler.

Fixed 2 problems:
* In ServletChannel, in COMPLETE state, there was a check comparing the number of bytes written with that declared by Content-Length.
Unfortunately the check was wrong in case of gzip, because it was comparing the application length with the gzipped length, resulting in a late sendErrorOrAbort() that was an abort() because the response was already fully sent.
Furthermore, if sendErrorOrAbort() was actually an abort(), there was an unnecessary attempt to complete the output.
* In GzipHandlerResponseAndCallback there were 2 last writes: one when the application calls output.close(), and one when the callback is succeeded.
Furthermore, when no content needs to be written, it was still compressed, causing an IOException in ChannelResponse (again when double-checking the bytes written, since Content-Length was 0, but the gzip header and trailer were written).

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lachlan-roberts commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java:
```diff
@@ -209,7 +209,7 @@ public ServletChannelState getServletRequestState()
     }
 
-    public long getBytesWritten()
+    private long getBytesWritten()
```
Is it okay to change this to private now?

joakime replied:
ServletChannel.getBytesWritten() is used by metrics libraries, it should remain public.

sbordet replied:
@joakime `ServletChannel.getBytesWritten()` should not be used by anyone, in general the whole `ServletChannel` class should be internal.
We have many other public ways to obtain the number of bytes written, so I think it's fine to make it private as it should have not been public in the first place.

lachlan-roberts commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -220,14 +222,23 @@ protected void commit(boolean last, Callback callback, ByteBuffer content)
+        // If there is nothing to write, don't compress.
+        if (last && BufferUtil.isEmpty(content))
+        {
+            LOG.debug("{} exclude by nothing to write", this);
```
Use `if (LOG.isDebugEnabled())`

sbordet replied:
Ok, but will change the whole class, which for some reason never used this pattern.

lachlan-roberts commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -220,14 +222,23 @@ protected void commit(boolean last, Callback callback, ByteBuffer content)
 
+        // We are varying the response due to accept encoding header.
+        if (_vary != null)
+            fields.ensureField(_vary);
```
What is the reason to move this here?
And why not put it before the other exclusions?

sbordet replied:
From [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html#field.vary):

> The "Vary" header field in a response describes what parts of a request message, aside from the method and target URI, might have influenced the origin server's process for selecting the content of this response.

So in theory we should always add it when we have the GzipHandler installed, as any response might change since it may be compressed.

However, this will break a number of tests.

Not sure if we interpreted the semantic wrong before (and wrote wrong tests) but I tried to make the minimum amount of changes in this PR.

However, I do think we should square this out once and for all.

@lachlan-roberts your thoughts?

sbordet replied:
@lachlan-roberts not sure if the interpretation could be:

If you ask the server for an empty response, it will always be empty and not compressed so it will be independent from the `Accept-Encoding` header, so it should not have the `Vary` header.

However, the RFC says "aside from the method and target URI".
With these 2 excluded, it basically becomes an announcement of the capabilities of the server, so it should always be there.

I'm not sure what exactly the purpose is, if the method and URI are not considered.

lachlan-roberts replied:
> If you ask the server for an empty response, it will always be empty and not compressed so it will be independent from the Accept-Encoding header, so it should not have the Vary header.

I can see how that would make sense. But the response might not always be empty so I think its better to always advertise the servers capability to vary the response on the `Accept-Encoding` header. Should probably just set it in `GzipHandler` instead of `GzipResponseAndCallback`.

Also what about the case where we skip compression because the `CONTENT_ENCODING` has already been set? right now we expect someone else to set the vary header.

lachlan-roberts reviewed (changes requested):
@sbordet there are tests failing, not sure if they are related.

sbordet committed:
Updates semantic of when to add the "Vary" header.

Now it is always only added by GzipHandler, when it would be possible to generate a response that might be compressed (even if it will not).

Added *.bz2 mime-type.
Updated GzipHandler with the correct mime-type for *.bz2: application/x-bzip2.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java:
```diff
@@ -544,14 +544,16 @@ public void handle()
                         {
-                            long written = getBytesWritten();
-                            if (getServletContextResponse().isContentIncomplete(written) && sendErrorOrAbort("Insufficient content written %d < %d".formatted(written, getServletContextResponse().getContentLength())))
+                            long written = getHttpOutput().getWritten();
```
We need a comment here saying why `getByteWritten()` is not used. i.e. because we want to look at the bytes written to the `HttpOutputStream` and not the bytes written to the underlying core response.

It would also be good to comment in `ServletContextResponse` that its `_contentLength` field is the length set by the application, but not necessarily the one set on the response as it may be removed/adjsuted.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -167,27 +164,27 @@ protected void commit(boolean last, Callback callback, ByteBuffer content)
-        // Are we excluding because of status?
         Request request = getRequest();
         Response response = this;
+        HttpFields.Mutable fields = response.getHeaders();
```
If we wanted to be a tiny bit more efficient here, we could iterate over all the fields once and extract `ETAG`, `CONTENT_TYPE`, `CONTENT_ENCODING`, and `CONTENT_LENGTH` in a single loop.

But perhaps the number of response fields is not large enough (and iterators cheap enough) that this is not worth it.

lorban replied:
I wrote a JMH benchmark, and iteration is so darn cheap that the extra cost of using the iterator takes a lot to amortize.

It takes a `ImmutableHttpFields` instance containing 20 fields to make looking up the last 4 fields of the internal array cheaper with the iterator.

Looking up less than 4 fields, or having to iterate less than 20 entries per lookup makes the solution using `get` always faster.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -95,10 +89,12 @@ public void succeeded()
-            _callback.succeeded();
+            // We need to write nothing here to intercept the committing of the
+            // response and possibly change headers in case write is never called.
+            if (_last)
+                _callback.succeeded();
+            else
+                write(true, null, _callback);
```
nice!

gregw reviewed (commented):
Just some comments for now until I see a clean build

sbordet committed:
Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12/11095/ise-gzip-handler

joakime reviewed (changes requested):
The test `testStaticContent(String)[3] ee10 – org.eclipse.jetty.tests.distribution.DemoModulesTests` has been consistently broken for the last 8 builds.

That doesn't smell like a flake.

sbordet committed:
Improved checks for the number of bytes written against declared content-length.
This is necessary because welcome files are written by DefaultServlet (most of the times) bypassing the Servlet classes, using ServletContextResponse directly.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'fix/jetty-12/11095/ise-gzip-handler'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw the test failure was due to welcome files bypassing the Servlet's OutputStream (using core classes directly), so that the check on the content-length was failing (as 0 bytes were written through the HttpOutput).
Please re-review.

gregw commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextResponse.java:
```diff
@@ -278,6 +281,18 @@ public long getContentLength()
+        super.write(last, content, callback);
+    }
+
+    public long getContentBytesWritten()
+    {
+        return _written;
+    }
```
> @gregw the test failure was due to welcome files bypassing the Servlet's OutputStream (using core classes directly), so that the check on the content-length was failing (as 0 bytes were written through the HttpOutput). Please re-review.

I'm not exactly how this method address this issue? The `getContentBytesWritten()` here looks like it will still be bypassed?  But fixed the tests... so I'm good... but an explanation would help.

Also, this method has the same name as `Response.getContentBytesWritten(Response)`, but returns a different value (i.e. pre compressed rather than post compressed).      Can you put some javadoc on both methods.   If you can think of a better name for this one, then change the name... maybe just `getWritten()` ?  Maybe make it package protected?

sbordet replied:
@gregw I made the method package local, added javadocs, but also filed #11325 because the issue needs to be investigated further, but it is orthogonal to GzipHandler.

lorban replied:
The bypassing in question is done by the `ResourceService` used in the `DefaultServlet` when a static file is served and `DefaultServlet` notices there is no wrapping. In that case, `DefaultServlet` passes the `ServletContextResponse` directly to `ResourceService` instead of wrapping the `HttpServletRequest` as a core `Request` and passing that to `ResourceService`.

The problem this creates here is that `HttpOutput` is unaware of what `ResourceService` writes, and `getBytesWritten` returns how many bytes were written to the socket after `GZIPHandler` compressed them.

So `ServletChannel.handle()` had no reliable way to know how many bytes were written to check if `Content-Length` is correct, hence that extra bytes counter in `ServletContextResponse` was added so that `ServletChannel.handle()` can make that `Content-Length` check.

gregw replied:
I'm thinking that the Servlet stuff only needs to track bytes written enough to do the auto commit/flush/close required by the spec.   Perhaps we are doing too much due to tests left over from Jetty-11 when this was the main API for response.    Let's consider this in #11325 for the next release.

gregw reviewed (dismissed):
Just some niggles at this stage, so conditional approval.

sbordet committed:
Improvements from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

minduch commented:
@sbordet Thank you very much!
When is (this fix available in) 12.0.6 [approximately] released?
Best regards, Christopher Mindus

sbordet commented:
@minduch early next week if testing goes fine.
